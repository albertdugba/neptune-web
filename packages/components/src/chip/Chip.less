@import (reference) '@transferwise/neptune-css/src/variables/neptune-tokens.less';

.chip {
  color: var(--color-text-primary);
  background: @color-base-smoke-light;
  margin-right: var(--size-4);
  height: var(--size-32);
  border-radius: 100px;

  &:active {
    background: @color-base-smoke-mid;
  }

  .chip-remove {
    opacity: 0.6;
    height: 100%;
  }

  &.has-error {
    background: var(--color-interactive-negative);
    color: @color-text-control;
  }
}
