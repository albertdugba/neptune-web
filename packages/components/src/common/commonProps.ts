export type CommonProps = {
  /**
   * Space-separated list of the case-sensitive CSS classes,
   * most would be applied to outermost element inside the component
   *
   * @see https://github.com/transferwise/neptune-web/blob/main/rfc/0001-always-pass-classname.md
   */
  className?: string;
};

export type AriaLabelProp = {
  'aria-label'?: string;
};

export type LinkProps = {
  href: string;
  target?: '_blank' | '_self' | '_parent' | '_top';
};
