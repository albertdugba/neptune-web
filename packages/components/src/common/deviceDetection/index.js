export { isIosDevice } from './deviceDetection';
export { isTouchDevice } from './isTouchDevice';
