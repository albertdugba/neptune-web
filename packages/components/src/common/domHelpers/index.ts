export function isServerSideRendering(): boolean {
  return typeof window === 'undefined';
}

export {
  addClickClassToDocumentOnIos,
  removeClickClassFromDocumentOnIos,
} from './documentIosClick';
