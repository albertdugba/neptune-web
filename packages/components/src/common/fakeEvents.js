export const fakeEvent = () => ({
  nativeEvent: { stopImmediatePropagation() {} },
  stopPropagation() {},
  preventDefault() {},
});

export const fakeKeyDownEventForKey = (keyCode) => ({
  ...fakeEvent(),
  keyCode,
});

export const fakeKeyDownEvent = (eventAttribute = {}) => ({
  ...fakeEvent(),
  ...eventAttribute,
});
