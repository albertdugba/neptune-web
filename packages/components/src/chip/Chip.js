import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Cross as CrossIcon } from '@transferwise/icons';
import './Chip.css';

import { useDirection } from '../common/hooks';

const Chip = (props) => {
  const { className, hasError, label, onRemove } = props;
  const { isRTL } = useDirection();

  return (
    <div
      className={classNames(
        'chip',
        'd-inline-flex',
        'align-items-center',
        'm-t-1',
        { 'has-error': hasError, 'p-r-1 p-l-2': !isRTL, 'p-l-1 p-r-2': isRTL },
        className,
      )}
    >
      {label}
      <button
        type="button"
        className={classNames('btn-unstyled chip-remove', { 'm-l-2': !isRTL, 'm-r-2': isRTL })}
        onClick={onRemove}
      >
        <CrossIcon />
      </button>
    </div>
  );
};

Chip.propTypes = {
  label: PropTypes.string.isRequired,
  onRemove: PropTypes.func.isRequired,
  className: PropTypes.string,
  hasError: PropTypes.bool,
};

Chip.defaultProps = {
  className: '',
  hasError: false,
};

export default Chip;
