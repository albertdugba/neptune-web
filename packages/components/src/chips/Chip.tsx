import classNames from 'classnames';
import { ReactElement } from 'react';

import { AriaLabelProperty, CommonProps } from '../common';
import CloseButton from '../common/closeButton/CloseButton';

export type ChipValue = string | number;

export type Props = {
  value?: ChipValue;
  label: string;
  onRemove?: () => void;
  onClick?: () => void;
  onKeyPress?: () => void;
  closeButton?: AriaLabelProperty;
  'aria-checked'?: boolean;
  role?: string;
} & AriaLabelProperty &
  CommonProps;

const Chip = ({
  label,
  value,
  onRemove,
  onClick,
  onKeyPress,
  className = undefined,
  'aria-label': ariaLabel,
  'aria-checked': ariaChecked,
  role,
  closeButton,
}: Props): ReactElement => {
  const isActionable = onClick || onKeyPress;
  const defaultRole = isActionable ? 'button' : undefined;
  const tabIndex = isActionable ? 0 : -1;
  return (
    <div
      key={value}
      role={role ?? defaultRole}
      tabIndex={tabIndex}
      aria-label={ariaLabel}
      aria-checked={ariaChecked}
      className={classNames(
        'np-chip',
        'd-flex',
        'align-items-center',
        'justify-content-between',
        className,
      )}
      {...(isActionable && { onClick, onKeyPress })}
    >
      <span aria-hidden={!!onRemove} className="np-chip-label">
        {label}
      </span>
      {onRemove ? (
        <CloseButton
          /* @ts-expect-error remove when CloseButton will be written on TS */
          className="btn-unstyled m-l-1"
          aria-label={closeButton && closeButton['aria-label']}
          filled
          size={16}
          onClick={onRemove}
        />
      ) : null}
    </div>
  );
};

export default Chip;
