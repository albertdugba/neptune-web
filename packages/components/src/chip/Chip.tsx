import { MouseEventHandler, ReactElement } from 'react';
import classNames from 'classnames';
import { Cross as CrossIcon } from '@transferwise/icons';
import { useDirection } from '../common/hooks';
import { CommonProps } from '../common';
import './Chip.css';

type Props = {
  /**
   * label for Chip
   */
  label: string;
  /**
   * onClick for on x
   */
  onRemove: MouseEventHandler;
  hasError?: boolean;
} & CommonProps;

/**
 * `Chip` component
 *
 * Docs link: https://transferwise.github.io/neptune-web/components/internal/Chip
 */
export default function Chip({
  label,
  onRemove,
  hasError = false,
  className = undefined,
}: Props): ReactElement {
  const { isRTL } = useDirection();
  return (
    <div
      className={classNames(
        'chip',
        'd-inline-flex',
        'align-items-center',
        'm-t-1',
        { 'has-error': hasError, 'p-r-1 p-l-2': !isRTL, 'p-l-1 p-r-2': isRTL },
        className,
      )}
    >
      {label}
      <button
        type="button"
        className={classNames('btn-unstyled chip-remove', { 'm-l-2': !isRTL, 'm-r-2': isRTL })}
        onClick={onRemove}
      >
        <CrossIcon />
      </button>
    </div>
  );
}
