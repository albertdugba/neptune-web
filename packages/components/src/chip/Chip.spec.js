import { shallow } from 'enzyme';

import { fakeEvent } from '../common/fakeEvents';
import { useDirection } from '../common/hooks';

import Chip from './Chip';

jest.mock('../common/hooks/useDirection');

describe('option', () => {
  let component;
  let props;

  const removeButton = () => component.find('.chip-remove');

  beforeEach(() => {
    props = {
      onRemove: jest.fn(),
      label: 'Test',
      hasError: false,
    };
    useDirection.mockImplementation(() => ({ direction: 'ltr', isRTL: false }));
    component = shallow(<Chip {...props} />);
  });

  it('renders', () => {
    expect(component.html()).toMatchSnapshot();
  });

  it('renders with RTL support', () => {
    useDirection.mockImplementation(() => ({ direction: 'rtl', isRTL: true }));
    component = shallow(<Chip {...props} />);
    expect(component.html()).toMatchSnapshot();
  });

  it('renders label', () => {
    expect(component.childAt(0).text()).toStrictEqual(props.label);
  });

  it('calls onRemove when remove button clicked', () => {
    removeButton().simulate('click', fakeEvent());
    expect(props.onRemove).toHaveBeenCalledTimes(1);
  });

  it('renders invalid chip correctly', () => {
    expect(component.hasClass('has-error')).toBe(false);
    component.setProps({
      hasError: true,
    });
    expect(component.hasClass('has-error')).toBe(true);
  });

  it('render with correct className', () => {
    expect(component.hasClass('test-class')).toBe(false);
    component.setProps({
      className: 'test-class',
    });
    expect(component.hasClass('test-class')).toBe(true);
  });
});
