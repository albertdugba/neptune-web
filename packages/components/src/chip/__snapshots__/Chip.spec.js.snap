// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`option renders 1`] = `"<div class=\\"chip d-inline-flex align-items-center m-t-1 p-r-1 p-l-2\\">Test<button type=\\"button\\" class=\\"btn-unstyled chip-remove m-l-2\\"><span class=\\"tw-icon tw-icon-cross \\" aria-hidden=\\"true\\" role=\\"presentation\\"><svg width=\\"16\\" height=\\"16\\" fill=\\"currentColor\\" focusable=\\"false\\"><path d=\\"M9.131 8l5.435-5.434-1.132-1.132L8 6.87 2.566 1.434 1.434 2.566 6.87 8l-5.435 5.434 1.132 1.132L8 9.13l5.434 5.435 1.132-1.132L9.13 8z\\"></path></svg></span></button></div>"`;

exports[`option renders with RTL support 1`] = `"<div class=\\"chip d-inline-flex align-items-center m-t-1 p-l-1 p-r-2\\">Test<button type=\\"button\\" class=\\"btn-unstyled chip-remove m-r-2\\"><span class=\\"tw-icon tw-icon-cross \\" aria-hidden=\\"true\\" role=\\"presentation\\"><svg width=\\"16\\" height=\\"16\\" fill=\\"currentColor\\" focusable=\\"false\\"><path d=\\"M9.131 8l5.435-5.434-1.132-1.132L8 6.87 2.566 1.434 1.434 2.566 6.87 8l-5.435 5.434 1.132 1.132L8 9.13l5.434 5.435 1.132-1.132L9.13 8z\\"></path></svg></span></button></div>"`;
